### ISSUE TITLE
[JENKINS-69352] Multiple concurrent builds cause severe memory spikes

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-15#6901]
<p>Since the update from 2.346.2 to 2.346.3, multiple builds running in parallel (either concurrent builds from the same pipeline or multiple builds from different pipelines) cause memory spikes that exceed the heap size of the JVM by a lot which leads to kubernetes killing the jenkins pod (OOMkilled).</p>

<p><b>Setup:</b></p>

<p>Kubernetes container with memory request+limit = 8gb</p>

<p>Jenkins -Xmx=4g</p>

<p> </p>

<p><b>Steps to reproduce:</b></p>
<ul>
	<li>Upgrade to 2.346.3</li>
	<li>Run multiple builds in parallel (they should take some minutes)</li>
	<li>In the jenkins pod, you can see that the memory of the jenkins process spikes within a couple of seconds</li>
	<li>As soon as the memory exceeds the memory limit of the container, kubernetes kills the pod. We tested the same scenario with a memory limit of 32gb and -Xmx=4g which leads to the issue taking longer to occur.</li>
</ul>

<p>Downgrading to 2.346.2 solved the issue.</p>

<p> </p>

<p><b>Could it be a plugin?</b></p>

<p>The exact same set of plugins works with 2.346.2.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dotstone">dotstone</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-69352" target="_blank">Multiple concurrent builds cause severe memory spikes</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Critical
<li><b>component(s)</b>: core
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2022-09-06T18:32:20+00:00
<li><b>votes</b>: 3
<li><b>watchers</b>: 8
<li><b>imported</b>: 2025-11-16
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Since the update from 2.346.2 to 2.346.3, multiple builds running in parallel (either concurrent builds from the same pipeline or multiple builds from different pipelines) cause memory spikes that exceed the heap size of the JVM by a lot which leads to kubernetes killing the jenkins pod (OOMkilled).</p>

<p><b>Setup:</b></p>

<p>Kubernetes container with memory request+limit = 8gb</p>

<p>Jenkins -Xmx=4g</p>

<p> </p>

<p><b>Steps to reproduce:</b></p>
<ul>
	<li>Upgrade to 2.346.3</li>
	<li>Run multiple builds in parallel (they should take some minutes)</li>
	<li>In the jenkins pod, you can see that the memory of the jenkins process spikes within a couple of seconds</li>
	<li>As soon as the memory exceeds the memory limit of the container, kubernetes kills the pod. We tested the same scenario with a memory limit of 32gb and -Xmx=4g which leads to the issue taking longer to occur.</li>
</ul>

<p>Downgrading to 2.346.2 solved the issue.</p>

<p> </p>

<p><b>Could it be a plugin?</b></p>

<p>The exact same set of plugins works with 2.346.2.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins: 2.346.3<br/>
OS: Linux - 5.4.0-58-generic
```
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>georg_blumenschein's <a href="https://issues.jenkins.io/browse/JENKINS-69352?focusedId=429502&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-429502">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Jumping in for <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dotstone" class="user-hover" rel="dotstone">dotstone</a> </p>

<p>Maybe the description is unclear on that one, the memory spikes should not come from the heap because it should be fixed at 4g as stated in the description, the problem is with Jenkins itself, not with the agents</p>

<p> </p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=basil" class="user-hover" rel="basil">basil</a> all plugins are used with latest, so also the workflow-cps plugin has a high enough version, but thanks for the suggestion <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p> </p>

<p>One thing that was noticed is that if running the same configuration (Same Jenkins image, same requests/limits/xmx), but smaller node size (t3.xlarge) on AWS, the problem seems to be gone, but I am not sure what that implies</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Jenkins: 2.346.3
OS: Linux - 5.4.196-108.356.amzn2.x86_64 </pre>
</div></div>
<p> </p>

<p> </p>
</pre>
</details>

</details>
<p>Jumping in for <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dotstone" class="user-hover" rel="dotstone">dotstone</a> </p>

<p>Maybe the description is unclear on that one, the memory spikes should not come from the heap because it should be fixed at 4g as stated in the description, the problem is with Jenkins itself, not with the agents</p>

<p> </p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=basil" class="user-hover" rel="basil">basil</a> all plugins are used with latest, so also the workflow-cps plugin has a high enough version, but thanks for the suggestion <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p> </p>

<p>One thing that was noticed is that if running the same configuration (Same Jenkins image, same requests/limits/xmx), but smaller node size (t3.xlarge) on AWS, the problem seems to be gone, but I am not sure what that implies</p>

<pre>
Jenkins: 2.346.3
OS: Linux - 5.4.196-108.356.amzn2.x86_64 </pre>
<p> </p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>basil's <a href="https://issues.jenkins.io/browse/JENKINS-69352?focusedId=429518&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-429518">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<blockquote><p>the problem is with Jenkins itself, not with the agents</p></blockquote>

<p>Then my previous question remains unanswered:</p>

<blockquote><p>If the controller JVM, does the same issue persist with the 2.346.2 Docker image but the 2.346.3 jenkins.war file (that would tell us whether the regression is in the Java code or in the environment delivered in the Docker image such as OS version, Java version, etc).</p></blockquote>

<p>This question also remains unanswered:</p>

<blockquote><p>And finally have you done any analysis of the heap dump to report what is using up the heap?</p></blockquote>
</pre>
</details>

</details>
<blockquote><p>the problem is with Jenkins itself, not with the agents</p></blockquote>

<p>Then my previous question remains unanswered:</p>

<blockquote><p>If the controller JVM, does the same issue persist with the 2.346.2 Docker image but the 2.346.3 jenkins.war file (that would tell us whether the regression is in the Java code or in the environment delivered in the Docker image such as OS version, Java version, etc).</p></blockquote>

<p>This question also remains unanswered:</p>

<blockquote><p>And finally have you done any analysis of the heap dump to report what is using up the heap?</p></blockquote>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>basil's <a href="https://issues.jenkins.io/browse/JENKINS-69352?focusedId=429537&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-429537">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>OpenJDK 11.0.16, which we started shipping in the 2.346.3 Docker image, supports cgroups v2. That might impact the memory usage characteristics of the process as described in <a href="https://developers.redhat.com/articles/2022/04/19/java-17-whats-new-openjdks-container-awareness#opinionated_configuration" class="external-link" target="_blank" rel="nofollow noopener">https://developers.redhat.com/articles/2022/04/19/java-17-whats-new-openjdks-container-awareness#opinionated_configuration</a>.</p>
</pre>
</details>

</details>
<p>OpenJDK 11.0.16, which we started shipping in the 2.346.3 Docker image, supports cgroups v2. That might impact the memory usage characteristics of the process as described in <a href="https://developers.redhat.com/articles/2022/04/19/java-17-whats-new-openjdks-container-awareness#opinionated_configuration" class="external-link" target="_blank" rel="nofollow noopener">https://developers.redhat.com/articles/2022/04/19/java-17-whats-new-openjdks-container-awareness#opinionated_configuration</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>basil's <a href="https://issues.jenkins.io/browse/JENKINS-69352?focusedId=429538&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-429538">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Can you try running with <tt>-Xlog:os+container=trace</tt> and <tt>-Xlog:gc=info</tt> and compare the os/container and GC logs between the working and failing version? This information would also be helpful (again, both on the working and failing version):</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>java -XshowSettings:system -version
java -Xlog:gc=info -version
kubectl describe deployment
jmcd $PID VM.info | grep -A13 '(cgroup)'
</pre>
</div></div>

<p>Also, are you using cgroups v1, cgroups v2, or the unified hierarchy?</p>

<p>In OpenJDK 11.0.16, the Kubernetes CPU/memory limits can now be observed by the JVM, which can impact things like the GC algorithm chosen by OpenJDK. Try focusing your investigation along these lines and see if you discover anything interesting.</p>
</pre>
</details>

</details>
<p>Can you try running with <tt>-Xlog:os+container=trace</tt> and <tt>-Xlog:gc=info</tt> and compare the os/container and GC logs between the working and failing version? This information would also be helpful (again, both on the working and failing version):</p>



```
java -XshowSettings:system -version
java -Xlog:gc=info -version
kubectl describe deployment
jmcd $PID VM.info | grep -A13 '(cgroup)'
```

<p>Also, are you using cgroups v1, cgroups v2, or the unified hierarchy?</p>

<p>In OpenJDK 11.0.16, the Kubernetes CPU/memory limits can now be observed by the JVM, which can impact things like the GC algorithm chosen by OpenJDK. Try focusing your investigation along these lines and see if you discover anything interesting.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>basil's <a href="https://issues.jenkins.io/browse/JENKINS-69352?focusedId=429539&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-429539">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=allan_burdajewicz" class="user-hover" rel="allan_burdajewicz">allan_burdajewicz</a> informed me about <a href="https://bugs.openjdk.org/browse/JDK-8292260" class="external-link" target="_blank" rel="nofollow noopener">JDK-8292260</a>, which led me to file <a href="https://github.com/jenkinsci/docker/issues/1441" class="external-link" target="_blank" rel="nofollow noopener">jenkinsci/docker#1441</a>. At the time of this writing, <a href="https://adoptium.net/" class="external-link" target="_blank" rel="nofollow noopener">https://adoptium.net</a> contains the following notice:</p>
<blockquote><p><b>18th of August 2022:</b> Upstream OpenJDK has released fixes to <a href="https://mail.openjdk.org/pipermail/jdk-updates-dev/2022-August/016460.html" class="external-link" target="_blank" rel="nofollow noopener">resolve a regression</a> in some of the July releases. We are working to release 11.0.16.1+1 and 17.0.4.1+1. As usual, you can continue to track progress <a href="https://github.com/adoptium/adoptium/issues/153" class="external-link" target="_blank" rel="nofollow noopener"><b>by platform</b></a> or <a href="https://github.com/adoptium/adoptium/issues/152" class="external-link" target="_blank" rel="nofollow noopener"><b>by detailed release checklist</b></a>.</p></blockquote>
<p>At the time of this writing the 11.0.16.1+1 and 17.0.4.1+1 build tarballs have been published, but the Docker images have not. I think as soon as the Docker images are published we will want to kick off Dependabot updating and backporting.</p>

<p>Observations:</p>
<ul>
	<li>We absolutely want this to be fixed in the next weekly release.</li>
	<li>I do not think we want to ship 2.361.1 until this is fixed.</li>
	<li>We may possibly want to issue a new 2.346.x release with this fix as well.</li>
</ul>


<p>If you think you are being impacted by <a href="https://bugs.openjdk.org/browse/JDK-8292260" class="external-link" target="_blank" rel="nofollow noopener">JDK-8292260</a> and need help proving or disproving that theory, please send me your <tt>hs_err_pid</tt> crash file and I can assist with your investigation and try to come up with a workaround.</p>

<p>CC <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dduportal" class="user-hover" rel="dduportal">dduportal</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=timja" class="user-hover" rel="timja">timja</a></p>
</pre>
</details>

</details>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=allan_burdajewicz" class="user-hover" rel="allan_burdajewicz">allan_burdajewicz</a> informed me about <a href="https://bugs.openjdk.org/browse/JDK-8292260" class="external-link" target="_blank" rel="nofollow noopener">JDK-8292260</a>, which led me to file <a href="https://github.com/jenkinsci/docker/issues/1441" class="external-link" target="_blank" rel="nofollow noopener">jenkinsci/docker#1441</a>. At the time of this writing, <a href="https://adoptium.net/" class="external-link" target="_blank" rel="nofollow noopener">https://adoptium.net</a> contains the following notice:</p>
<blockquote><p><b>18th of August 2022:</b> Upstream OpenJDK has released fixes to <a href="https://mail.openjdk.org/pipermail/jdk-updates-dev/2022-August/016460.html" class="external-link" target="_blank" rel="nofollow noopener">resolve a regression</a> in some of the July releases. We are working to release 11.0.16.1+1 and 17.0.4.1+1. As usual, you can continue to track progress <a href="https://github.com/adoptium/adoptium/issues/153" class="external-link" target="_blank" rel="nofollow noopener"><b>by platform</b></a> or <a href="https://github.com/adoptium/adoptium/issues/152" class="external-link" target="_blank" rel="nofollow noopener"><b>by detailed release checklist</b></a>.</p></blockquote>
<p>At the time of this writing the 11.0.16.1+1 and 17.0.4.1+1 build tarballs have been published, but the Docker images have not. I think as soon as the Docker images are published we will want to kick off Dependabot updating and backporting.</p>

<p>Observations:</p>
<ul>
	<li>We absolutely want this to be fixed in the next weekly release.</li>
	<li>I do not think we want to ship 2.361.1 until this is fixed.</li>
	<li>We may possibly want to issue a new 2.346.x release with this fix as well.</li>
</ul>


<p>If you think you are being impacted by <a href="https://bugs.openjdk.org/browse/JDK-8292260" class="external-link" target="_blank" rel="nofollow noopener">JDK-8292260</a> and need help proving or disproving that theory, please send me your <tt>hs_err_pid</tt> crash file and I can assist with your investigation and try to come up with a workaround.</p>

<p>CC <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dduportal" class="user-hover" rel="dduportal">dduportal</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=timja" class="user-hover" rel="timja">timja</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>basil's <a href="https://issues.jenkins.io/browse/JENKINS-69352?focusedId=429985&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-429985">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Although we never got proof (in the form of an <tt>hs_err</tt> log) that this was the C2 JIT issue, I think it is quite likely that it was and this has been fixed in 2.346.3-2. If this issue occurs again, please open a new ticket.</p>
</pre>
</details>

</details>
<p>Although we never got proof (in the form of an <tt>hs_err</tt> log) that this was the C2 JIT issue, I think it is quite likely that it was and this has been fixed in 2.346.3-2. If this issue occurs again, please open a new ticket.</p>
